File: backend/src/main/kotlin/org/genspectrum/dashboardsbackend/api/ApiKey.kt
package org.genspectrum.dashboardsbackend.api

import kotlin.time.Instant

/** Public metadata about a user's API key. The raw key is never returned after creation. */
data class ApiKeyMetadata(
    val createdAt: Instant,
    val lastUsedAt: Instant?,
)

/** Returned once when a key is first generated. The raw key is never stored and cannot be retrieved again. */
data class GeneratedApiKey(
    val key: String,
    val createdAt: Instant,
)

/** Request body for the internal validate endpoint called by the proxy. */
data class ValidateApiKeyRequest(val key: String)

/** Response from the internal validate endpoint — contains the internal user ID for the validated key. */
data class ValidateApiKeyResponse(val userId: Long)

File: website/src/types/ApiKey.ts
import { z } from 'zod';

export const apiKeyMetadataSchema = z.object({
    createdAt: z.string(),
    lastUsedAt: z.string().nullable(),
});

export type ApiKeyMetadata = z.infer<typeof apiKeyMetadataSchema>;

export const generatedApiKeySchema = z.object({
    key: z.string(),
    createdAt: z.string(),
});

export type GeneratedApiKey = z.infer<typeof generatedApiKeySchema>;
